specification/dsdl/architecture.tex

```diff
@@ -189,25 +198,23 @@ \subsection{Elements of data type definition}
 As explained above, one data type definition contains either one or two data schema definitions,
 for message types and service types, respectively.
 
-The term \emph{attribute} is used to refer to either a field (including padding fields) or a constant
-associated with a particular object or type.
-
-A data schema definition contains an ordered, possibly empty collection of \emph{fields} and/or
-unordered, possibly empty collection of \emph{constants}.
+A data schema definition contains an ordered, possibly empty collection of \emph{field attributes} and/or
+unordered, possibly empty collection of \emph{constant attributes}.
 A data schema may describe either a \emph{structure object} or a \emph{tagged union object}.
-The value of a structure object is a function of the values of all of its fields.
-A tagged union object is formed from at least two fields,
-but it is capable of holding exactly one field at any given time.
-The value of a tagged union object is a function of which field it is holding at the moment and the value of said field.
+The value of a structure object is a function of the values of all of its field attributes.
+A tagged union object is formed from at least two field attributes,
+but it is capable of holding exactly one field attribute value at any given time.
+The value of a tagged union object is a function of which field attribute value
```

Comment:
(suggested replacement for this sentence)
A tagged union's value is evaluated as a function of the tagged type the value is accessed through. There is no _intrinsic_ memo of an "actual" type associated with a union. Any such concept may, optionally, be provided by a type author using additional fields in a structure object that contains the tagged union.(footnote)

footnote - In this manner the DSDL tagged union mirrors the C language's concept of unions.

Reply:
> A tagged union's value is evaluated as a function of the tagged type the value is accessed through.

This would imply that the type of each option in a tagged union must be unique within that union. The following is valid according to the current draft, but not valid according to your proposed wording:

```python
@union
uint8 a
uint8 b
```

> There is no intrinsic memo of an "actual" type associated with a union.

The meaning of this sentence evades my understanding. Is it saying that unions are not types themselves?

Reply:
I am going to merge this now without your suggested change because I want to avoid merge conflicts in the target branch I'm currently editing. If you believe that this issue should be discussed further, please reply here or open a new ticket.

Reply:
I'm being confused about the definition of `value` in this paragraph. In your mind, is the union's value the state of all the union's bits in memory (including undefined bits where a union was assigned using a field whose size is < the total size of the union) or is the value the representation obtained by accessing a union via a given field (i.e. ignoring undefined bits)?

Reply:
This is a more abstract statement than either of your options. "Value" of the union here is an abstract entity that affects the behavior or state of the process or logic it is used in. The implication of my definition, when considered in this setting, is that other fields of the union (that is, all fields except the active one) are irrelevant and do not affect the "behavior or state of the process or logic" which uses it. I avoided using any more specific definition because this is supposed to be a very abstract section. Should it be reworked?

Reply:
meh. I think I might be bike-shedding now but I just don't like the use of the term "holding" (e.g. "...is a function of which field attribute value it is holding...") since that suggests some hidden mechanics that don't exist (i.e. "holding" is an active verb but unions are passive data constructs.

So, I'm okay with this I guess.

Reply:
We could replace "a function of which field attribute value it is holding" with "a function of which field attribute value is currently selected"? Or something like this. There is going to be another pass over the entire document when it's ready to weed out bad English and poorly phrased sentences (I'm not sure if Hamish is still available for this work; if not, we'll hire someone else).

Reply:
Agreed. We'll leave this stuff for the tech writer.